### ISSUE TITLE
Obtain a list of mapped axes

### ISSUE THREAD

**Issue by hamzamerzic** [google/jax#4191]
Is there a way to obtain a list of mapped axes from within a function? 

My use case is the following. I have a function where I would like to `jax.lax.pmean` over **all** mapped axes, so that I can run it with either `vmap` or `pmap` + `vmap`, or another depth of mapping. Being able to see the mapped axes and apply `pmean` in a custom order would be a great feature. It could also allow for better runtime error checking and debugging.

**Comment by apaszke** [comment]
We don't support a function like that at the moment, although this information is tracked internally. I think that the recommended way to do this would be to pass in the axes you want to reduce over as an argument to your function. Note that all reduction collectives accept tuples of axis names as well.

**Comment by hamzamerzic** [comment]
Thanks Adam. Do you foresee these internals being exposed to the user?

We are building a modular codebase, with functions arbitrarily nested, so following the current recommended way would require passing a list of mapped axes to every function. 

Also, since we work with Haiku, we have separate init and apply methods with shared initialization. Having a way to programatically obtain the list of mapped axes would simplify our code since during init we wouldn't need to map over all axes. Currently, with omnistaging, we have nested `pmap(vmap(fn, 'j'), 'i')`, and downstream modules reduce over `i` and `j`. Thus, on init we need to create a "fake" outer pmap to get the params on a single device - doing this multiple times makes the code overly verbose and impacts readability.

**Comment by mattjj** [comment]
Thanks for the question! (Hamza and I also talked about this out-of-band back when he first asked, but I'm only now getting around to replying here.)

We can't provide an API to do this because that would mean a user wouldn't be able to map a function and predict its behavior, since the function would be able to switch its behavior based on the mapping context. That is, it'd break mapping semantics. (One way to view this is it'd be a transfer of power, or complexity, from the user/caller to the library-writer/callee.)

Plumbing axis names yourself in Python is a way to resolve the potential ambiguity about what a function might do with mapped axes (and to make the caller and callee cooperate/power-share explicitly via an API). That is, the callee can specify its behavior explicitly in terms of what axis names are passed in, and the caller can control what axes it wants the callee to know about. Explicit is good!

Hope that provides a somewhat-satisfying rationale. I think as we improve the mapping constructs in JAX, pain points like this might fade away.
